### ISSUE TITLE
macro '...' must be defined before this point but is defined later

### ISSUE THREAD

**Issue by waterlink** [crystal-lang/crystal#2457]
This confuses me. What should I do and how to fix that?

Code demonstrating the problem on current crystal-lang/crystal@master can be found here https://github.com/waterlink/spec2.cr/blob/issues-with-new-compiler/issue_macro_must_be_defined_before.cr

To reproduce:

- `git clone https://github.com/waterlink/spec2.cr -b issues-with-new-compiler`
- `shards install --production`
- `crystal run issue_macro_must_be_defined_before.cr`

Crystal was compiled with the master branch pointing at this commit: https://github.com/crystal-lang/crystal/commit/0cf3ed9e7a6ea3023f05c5cd16a60350713a819d

**Comment by asterite** [comment]
The simplest code to understand the error is this:

```crystal
class Bar
  foo
end

macro foo
end
```

The compiler has several passes, and in all of them if a call resolves to a macro then it is expanded and semantic from that phase is applied. But if the macro is not found in the first phase (like in the example above) but is found in a later phase, it's an error.

If we change the error to "undefined macro '...'" is the error clearer? It's basically that something resolves to a macro that is defined at a point that comes after its use. The solution is usually to `require` the code that defines the macro before the macro call.

**Comment by waterlink** [comment]
@asterite What about macros, that generate another macros?

**Comment by waterlink** [comment]
Example:

```crystal
macro root_macro
  macro inner_macro
    ".. do stuff .."
  end
end
```

**Comment by waterlink** [comment]
Results in `can't declare macro dynamically` :( This was one of the key features, that were powering `mocks.cr`..

**Comment by waterlink** [comment]
Oh not. I am actually wrong on that one. This example works.

**Comment by waterlink** [comment]
@asterite  I will try to fiddle with requires then.

**Comment by waterlink** [comment]
Adding the require statement before main `spec2` require does not help:

```crystal
require "./src/matchers/raise_error"
require "./src/spec2"

Spec2.describe "an issue" do
  it "fails with 'macro must be defined before'" do
    expect do
      42
    end.to raise_error(Exception)
  end
end
```

`./src/matchers/raise_error` looks like:

```crystal
module Spec2
  module Matchers
    macro raise_error(klass=Exception, message=nil)
      ...
    end
  end
end
```

And `Spec2::Matchers` gets inherited from the `module Spec2::DSL`, that is included in root context, and then every context and example inherits from this root context directly or indirectly.

**Comment by waterlink** [comment]
Fully macro-expanded code (at `0.15.0`) looks like that for the example above:

```crystal
__temp_91 = @@__spec2_active_context
module Spec2__AnIssue
  include ::Spec2::DSL::Spec2___
  SPEC2_FULL_CONTEXT = ":root -> an issue (issue_macro_must_be_defined_before.cr:4)"
  SPEC2_CONTEXT = Spec2__AnIssue
  LETS = {} of String => Int32
  LETS_BANG = {} of String => Int32
  BEFORES = [] of Int32
  AFTERS = [] of Int32
  ITS = {"fails with 'macro must be defined before'" => do
    expect do
      42
    end.to(raise_error(Exception))
  end} of String => Int32
  @@__spec2_active_context = ::Spec2::Context.new(__temp_91, "an issue")
  (  __temp_91
  ::Spec2::Context.instance
).contexts << (@@__spec2_active_context)
  def __spec2_run_lets!
    super
  end
  def __spec2_clear_lets
    super
  end
  def __spec2_before_hook
    super
  end
  def __spec2_after_hook
    super
  end
  class Spec2__FailsWithMacroMustBeDefinedBefore < ::Spec2::Example
    include Spec2__AnIssue
    def initialize(context)
      @context = context
      @what = "fails with 'macro must be defined before'"
      @blk = -> do
      end
    end
    def run
      begin
        __spec2_delayed = [] of -> 
        __spec2_before_hook
        __spec2_run_lets!
        expect do
          42
        end.to(raise_error(Exception))
      ensure
        __spec2_after_hook
        __spec2_delayed.not_nil!.each(&.call)
      end
    end
  end
  __temp_92 = if __temp_231 = @@__spec2_active_context
    __temp_231
  else
    ::Spec2::Context.instance
  end
  __temp_92.examples << (Spec2__FailsWithMacroMustBeDefinedBefore.new(__temp_92))
end
```

**Comment by waterlink** [comment]
This is not a valid crystal source, rather a stringual representation of expanded AST tree

**Comment by waterlink** [comment]
And `::Spec2::DSL::Spec2___` is a module, that includes `::Spec2::DSL` and therefore includes/extends `::Spec2::Matchers`

**Comment by waterlink** [comment]
Maybe the problem is because it is both `include`ed and `extend`ed.. Will check that out.

**Comment by asterite** [comment]
When is the module that defines the macro included in the created types? I have this and it fails in a similar way:

```crystal
module Matcher
  macro raise_error
    p 1
  end
end

macro foo
  class Foo
    {{yield}}
  end
end

foo do
  # Macro is not found yet
  raise_error
end

class Foo
  include Matcher
end

Foo.new.bar
```

**Comment by asterite** [comment]
You'll have to reduce and extract the code much more for us to be able to understand the problem.

**Comment by waterlink** [comment]
That is exactly what I am doing.

I am using `{{block.body}}` all over the place.

**Comment by waterlink** [comment]
And I want these macro to be available for these blocks to re-use inside. Because I need a nested structure there..

**Comment by waterlink** [comment]
On the other hand, `include` in my case still should happen first.

**Comment by waterlink** [comment]
Additionally, expansion of `it`s is lazy, it happens only after all its are "defined" in the current context and stored to a compile-time-only constant, then they are iterated over and defined as classes.

**Comment by waterlink** [comment]
Ok, it is related to this laziness.

@asterite I was able to minimize the code:

```crystal
macro foo
  "foobar"
end

ITS = {} of String => Int32

macro it(desc, &blk)
  {% ITS[desc] = blk %}
end

it "works" do
  foo
end
```

Rationale behind doing that is to be able to define all the different `its`, `lets`, `befores` and `afters` before starting definition of any example classes, so that they can refer to all of them at that time. And only after the `context`'s block is fully defined, then there comes `def_lets`, `def_befores`, `def_afters` and `def_its`, that actually define these elements in correct order, so that example will have proper access to them and will actually compile.

**Comment by asterite** [comment]
Hmm... looks like a compiler bug to me, 99% sure about that, I'll investigate.

**Comment by asterite** [comment]
Mmm... actually, it's because of this:

1. First pass, the compiler expands the macro "it", so ITS becomes `{"foo" => foo}`
2. Second pass (type guess), the compiler visits the constant and it's `ITS => {"foo", foo}`, so `foo` is visited, the macro is expanded but it didn't expand to a macro before (that node was never visited), so it gives an error.

I can fix this by avoiding visiting constant declarations in this pass (maybe). But how do you use `ITS` later?

**Comment by asterite** [comment]
I think it will fail anyone because if you "paste" `foo` later, it will resolve to the macro call but it wasn't analyzed in a previous pass. If `foo` for example declares a class then that could be problematic because that pass already passed (I don't know if this is the case, but in general it's a problematic solution)

**Comment by waterlink** [comment]
In context macro I have these lines:

```crystal
macro context(desc, &block)
   ...

  {{block.body}}
  def_its
end
```

where `def_its` is another macro, that is roughly this:

```crystal
macro def_its
  {% for desc in ITS %}
    {% blk = ITS[what] %}
    class AGeneratedExampleNameHere < ::Spec2::Example
      include {{SPEC2_CONTEXT}}

      def run
        .. do some setup ..
        {{blk.body}}
        .. do teardown
      end
    end
  {% end %}
end
```

**Comment by asterite** [comment]
@waterlink Don't worry. I'm sure this was working before and now it broke, but there's no reason for it to break. The only "bad thing" that can happen is if a class or method is defined in a pass that's not the first one (when classes and methods are defined). So I'll change the compiler to error on that case, but still allow macros to be expanded in other phases if they don't declare types/methods (as in your case).

**Comment by waterlink** [comment]
@asterite We could try that. I am pretty sure that would work in my case. Thanks!

**Comment by asterite** [comment]
@waterlink Can you try to build a compiler from [this branch](https://github.com/crystal-lang/crystal/tree/feature/fix_2457) and see if this fixes it?

**Comment by waterlink** [comment]
That fixes this concrete problems. Now I just have normal `can't infer type of @something_something`, that I will be fixing now. Thanks!
